File: rfid/full_workload.c
#include <stdio.h>

#define AMPLITUDE 127  // Amplitude of the sine wave
#define SAMPLE_RATE 6  // in Hz
#define FREQUENCY_1 1  // Frequency for '1'
#define FREQUENCY_2 2  // Frequency for '0'
#define SYMBOL_DURATION 1  // Duration of each symbol in seconds
#define USER_ID "11110000"  // Example reader query
#define VERIFICATION "11000101"

// Precomputed sine wave lookup table (scaled to 8-bit integer range)
int sine_wave_lookup[360] = {
      0,    2,    4,    6,    8,   11,   13,   15,   17,   19,
     22,   24,   26,   28,   30,   32,   35,   37,   39,   41,
     43,   45,   47,   49,   51,   53,   55,   57,   59,   61,
     63,   65,   67,   69,   71,   72,   74,   76,   78,   79,
     81,   83,   84,   86,   88,   89,   91,   92,   94,   95,
     97,   98,  100,  101,  102,  104,  105,  106,  107,  108,
    109,  111,  112,  113,  114,  115,  116,  116,  117,  118,
    119,  120,  120,  121,  122,  122,  123,  123,  124,  124,
    125,  125,  125,  126,  126,  126,  126,  126,  126,  126,
    127,  126,  126,  126,  126,  126,  126,  126,  125,  125,
    125,  124,  124,  123,  123,  122,  122,  121,  120,  120,
    119,  118,  117,  116,  116,  115,  114,  113,  112,  111,
    109,  108,  107,  106,  105,  104,  102,  101,  100,   98,
     97,   95,   94,   92,   91,   89,   88,   86,   84,   83,
     81,   79,   78,   76,   74,   72,   71,   69,   67,   65,
     63,   61,   59,   57,   55,   53,   51,   49,   47,   45,
     43,   41,   39,   37,   35,   32,   30,   28,   26,   24,
     22,   19,   17,   15,   13,   11,    8,    6,    4,    2,
      0,   -2,   -4,   -6,   -8,  -11,  -13,  -15,  -17,  -19,
    -22,  -24,  -26,  -28,  -30,  -32,  -35,  -37,  -39,  -41,
    -43,  -45,  -47,  -49,  -51,  -53,  -55,  -57,  -59,  -61,
    -63,  -65,  -67,  -69,  -71,  -72,  -74,  -76,  -78,  -79,
    -81,  -83,  -84,  -86,  -88,  -89,  -91,  -92,  -94,  -95,
    -97,  -98, -100, -101, -102, -104, -105, -106, -107, -108,
   -109, -111, -112, -113, -114, -115, -116, -116, -117, -118,
   -119, -120, -120, -121, -122, -122, -123, -123, -124, -124,
   -125, -125, -125, -126, -126, -126, -126, -126, -126, -126,
   -127, -126, -126, -126, -126, -126, -126, -126, -125, -125,
   -125, -124, -124, -123, -123, -122, -122, -121, -120, -120,
   -119, -118, -117, -116, -116, -115, -114, -113, -112, -111,
   -109, -108, -107, -106, -105, -104, -102, -101, -100,  -98,
    -97,  -95,  -94,  -92,  -91,  -89,  -88,  -86,  -84,  -83,
    -81,  -79,  -78,  -76,  -74,  -72,  -71,  -69,  -67,  -65,
    -63,  -61,  -59,  -57,  -55,  -53,  -51,  -49,  -47,  -45,
    -43,  -41,  -39,  -37,  -35,  -32,  -30,  -28,  -26,  -24,
    -22,  -19,  -17,  -15,  -13,  -11,   -8,   -6,   -4,   -2,
};

// Helper function to calculate sine using a lookup table
int sine_lookup(int angle) {
    angle = angle % 360;
    return sine_wave_lookup[angle];
}

// Helper function to generate sine waves as integers
void generate_sine_wave(int* wave, int freq, int duration, int sample_rate, int phase) {
    int N = duration * sample_rate;
    for (int i = 0; i < N; i++) {
        int angle = ((360 * freq * i / sample_rate) + phase) % 360;
        wave[i] = sine_lookup(angle);
    }
}

// Helper function to compute the length of a string
int string_length(const char* str) {
    int length = 0;
    while (str[length] != '\0') {
        length++;
    }
    return length;
}

// Helper function to compare two strings
int string_compare(const char* str1, const char* str2) {
    while (*str1 && (*str1 == *str2)) {
        str1++;
        str2++;
    }
    return *(unsigned char*)str1 - *(unsigned char*)str2;
}

// Simple DFT function (not an optimized FFT) with integer math
void discrete_fourier_transform(int* signal, int N, int* real, int* imag) {
    for (int n = 0; n < N / 2; n++) {
        real[n] = 0;
        imag[n] = 0;
        for (int k = 0; k < N; k++) {
            int angle = (360 * n * k / N) % 360;
            real[n] += signal[k] * sine_lookup(360 - angle);
            imag[n] += signal[k] * sine_lookup(angle);
        }
    }
}

// Generate a sample FSK signal for a given binary sequence
void modulate_binary_sequence(const char* binary_sequence, int* fsk_signal, int frequency_1, int frequency_2, int symbol_duration, int sample_rate) {
    int samples_per_symbol = symbol_duration * sample_rate;
    int wave[samples_per_symbol];
    int index = 0;
    for (int i = 0; binary_sequence[i] != '\0'; i++) {
        int freq = (binary_sequence[i] == '1') ? frequency_1 : frequency_2;
        generate_sine_wave(wave, freq, symbol_duration, sample_rate, 0);
        for (int j = 0; j < samples_per_symbol; j++) {
            fsk_signal[index++] = wave[j];
        }
    }
}

// Demodulate FSK signal to binary sequence
void demodulate_fsk_signal(int* fsk_signal, char* decoded_bits, int frequency_1, int frequency_2, int symbol_duration, int sample_rate, int len) {
    int samples_per_symbol = symbol_duration * sample_rate;
    int real[samples_per_symbol / 2], imag[samples_per_symbol / 2];
    for (int i = 0; i < len; i += samples_per_symbol) {
        discrete_fourier_transform(fsk_signal + i, samples_per_symbol, real, imag);
        int index_1 = frequency_1 * samples_per_symbol / sample_rate;
        int index_2 = frequency_2 * samples_per_symbol / sample_rate;
        int power_1 = real[index_1] * real[index_1] + imag[index_1] * imag[index_1];
        int power_2 = real[index_2] * real[index_2] + imag[index_2] * imag[index_2];
        decoded_bits[i / samples_per_symbol] = (power_1 > power_2) ? '1' : '0';
    }
    decoded_bits[len / samples_per_symbol] = '\0';
}

int main() {
    // Test data generation
    const char* binary_sequence = "11000101";
    int signal_length = string_length(binary_sequence) * SYMBOL_DURATION * SAMPLE_RATE;
    int generated_signal[signal_length];
    modulate_binary_sequence(binary_sequence, generated_signal, FREQUENCY_1, FREQUENCY_2, SYMBOL_DURATION, SAMPLE_RATE);

    // Simulate receiving and demodulating the signal at the reader
    char decoded_sequence[string_length(binary_sequence) + 1];
    demodulate_fsk_signal(generated_signal, decoded_sequence, FREQUENCY_1, FREQUENCY_2, SYMBOL_DURATION, SAMPLE_RATE, signal_length);

    // Gate the signal based on the reader's query
    int output_signal[signal_length];
    if (string_compare(VERIFICATION, decoded_sequence) == 0) {
        modulate_binary_sequence(USER_ID, output_signal, FREQUENCY_1, FREQUENCY_2, SYMBOL_DURATION, SAMPLE_RATE);
    } else {
        for (int i = 0; i < signal_length; i++) {
            output_signal[i] = 0;
        }
    }

    // Modulate the output signal to send out
    char check_output[string_length(binary_sequence) + 1];
    demodulate_fsk_signal(output_signal, check_output, FREQUENCY_1, FREQUENCY_2, SYMBOL_DURATION, SAMPLE_RATE, signal_length);

    // Outputs for verification
    printf("Original Binary Sequence: %s\n", binary_sequence);
    printf("Demodulated Binary Sequence on tag: %s\n", decoded_sequence);
    printf("Transmitter received Binary Sequence: %s\n", check_output);

    return 0;
}

File: rfid/full_workload.py
import math

# Define constants for scaling
AMPLITUDE = 127  # Amplitude of the sine wave
SAMPLE_RATE = 6  # in Hz
FREQUENCY_1 = 1  # Frequency for '1'
FREQUENCY_2 = 2  # Frequency for '0'
SYMBOL_DURATION = 1  # Duration of each symbol in seconds
USER_ID = '11110000'  # Example reader query, expecting this to be the correct signal to respond to
VERIFICATION = '11000101'

# Helper function to generate sine waves as integers
def generate_sine_wave(freq, duration, sample_rate, phase=0):
    return [int(AMPLITUDE * math.sin(2 * math.pi * freq * (i / sample_rate) + phase)) for i in range(int(duration * sample_rate))]

# Simple DFT function (not an optimized FFT) with integer math
def discrete_fourier_transform(signal, sample_rate):
    N = len(signal)
    dft_result = []
    for n in range(N // 2):
        real = sum(signal[k] * int(AMPLITUDE * math.cos(-2 * math.pi * n * k / N)) for k in range(N))
        imag = sum(signal[k] * int(AMPLITUDE * math.sin(-2 * math.pi * n * k / N)) for k in range(N))
        dft_result.append(complex(real, imag))
    return dft_result

# Generate a sample FSK signal for a given binary sequence
def modulate_binary_sequence(binary_sequence, frequency_1, frequency_2, symbol_duration, sample_rate):
    fsk_signal = []
    for bit in binary_sequence:
        if bit == '1':
            fsk_signal.extend(generate_sine_wave(frequency_1, symbol_duration, sample_rate))
        else:
            fsk_signal.extend(generate_sine_wave(frequency_2, symbol_duration, sample_rate))
    return fsk_signal

# Demodulate FSK signal to binary sequence
def demodulate_fsk_signal(fsk_signal, frequency_1, frequency_2, symbol_duration, sample_rate):
    samples_per_symbol = int(symbol_duration * sample_rate)
    decoded_bits = []
    for i in range(0, len(fsk_signal), samples_per_symbol):
        symbol_data = fsk_signal[i:i+samples_per_symbol]
        dft_result = discrete_fourier_transform(symbol_data, sample_rate)
        index_1 = int(frequency_1 * samples_per_symbol / sample_rate)
        index_2 = int(frequency_2 * samples_per_symbol / sample_rate)
        power_1 = abs(dft_result[index_1]) ** 2
        power_2 = abs(dft_result[index_2]) ** 2
        decoded_bits.append('1' if power_1 > power_2 else '0')
    return ''.join(decoded_bits)

# Invert binary sequence
def invert_bits(binary_sequence):
    return ''.join('1' if x == '0' else '0' for x in binary_sequence)

# Test data generation
binary_sequence = '11000101'  # RFID Tag's data
generated_signal = modulate_binary_sequence(binary_sequence, FREQUENCY_1, FREQUENCY_2, SYMBOL_DURATION, SAMPLE_RATE)

# Simulate receiving and demodulating the signal at the reader
decoded_sequence = demodulate_fsk_signal(generated_signal, FREQUENCY_1, FREQUENCY_2, SYMBOL_DURATION, SAMPLE_RATE)

# Gate the signal based on the reader's query
output_signal = []
if VERIFICATION == decoded_sequence:
    output_signal = modulate_binary_sequence(USER_ID, FREQUENCY_1, FREQUENCY_2, SYMBOL_DURATION, SAMPLE_RATE)
else:
    output_signal = [0] * len(generated_signal)  # The tag gates the signal, sending nothing

# Modulate the output signal to send out
check_output = demodulate_fsk_signal(output_signal, FREQUENCY_1, FREQUENCY_2, SYMBOL_DURATION, SAMPLE_RATE)

# Outputs for verification
print("Original Binary Sequence:", binary_sequence)
print("Decoded Binary Sequence:", decoded_sequence)
print("Inverted Binary Sequence:", check_output)
